src/kani-verifier/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+fn download_url() -> String {
+    let tag: &str = &format!("kani-{}", VERSION);
+    let file: &str = &download_filename();
+    format!("https://github.com/model-checking/kani/releases/download/{}/{}", tag, file)
+}
+
+/// Effectively the entry point (i.e. `main` function) for both our proxy binaries.
+pub fn proxy(bin: &str) -> Result<()> {
+    // In an effort to keep our dependencies minimal, we do the bare minimum argument parsing
+    let args: Vec<_> = std::env::args_os().collect();
+    if args.len() >= 2 && args[1] == "setup" {
+        if args.len() >= 4 && args[2] == "--use-local-bundle" {
+            setup(Some(args[3].clone()))
+        } else {
+            setup(None)
+        }
+    } else {
+        if !appears_setup() {
+            setup(None)?;
+        }
+        exec(bin)
+    }
+}
+
+/// Fast check to see if we look setup already
+fn appears_setup() -> bool {
+    kani_dir().exists()
+}
+
+/// Sets up Kani by unpacking/installing to `~/.kani/kani-VERSION`
+fn setup(use_local_bundle: Option<OsString>) -> Result<()> {
+    // e.g. `~/.kani/`
+    let kani_dir = kani_dir();
+    let base_dir = kani_dir.parent().expect("No base directory?");
+
+    println!("Running Kani first-time setup...");
+
+    // 0. Make sure we have a `~/.kani/`
+    std::fs::create_dir_all(&base_dir)?;
+
+    // 1. Download and unpack our release binary
+    if let Some(pathstr) = use_local_bundle {
+        let path = Path::new(&pathstr).canonicalize()?;
+        Command::new("tar").arg("zxf").arg(&path).current_dir(base_dir).run()?;
+
+        // when given a bundle, it's often "-latest" but we expect "-1.0" or something. Hack it up.
+        let file = path.file_name().expect("has filename").to_string_lossy();
+        let components: Vec<_> = file.split("-").collect();
+        let expected_dir = format!("{}-{}", components[0], components[1]);
+
+        std::fs::rename(base_dir.join(expected_dir), &kani_dir)?;
+    } else {
+        let bundle = base_dir.join(download_filename());
+        Command::new("curl").args(&["-sSLf", "-o"]).arg(&bundle).arg(download_url()).run()?;
+
+        Command::new("tar").arg("zxf").arg(&bundle).current_dir(base_dir).run()?;
+
+        std::fs::remove_file(bundle)?;
+    }
+
+    // 2. We need a specific rustup toolchain
+    let toolchain_version = std::fs::read_to_string(kani_dir.join("rust-toolchain"))?;
+    Command::new("rustup").args(&["toolchain", "install", &toolchain_version]).run()?;
+
+    let toolchain = home::rustup_home()?.join("toolchains").join(toolchain_version);
+
+    // 3. Connect to our toolchain
+    Command::new("ln").arg("-s").arg(toolchain).arg(kani_dir.join("toolchain")).run()?;
+
+    // 4. Install our python dependencies
+    let pyroot = kani_dir.join("pyroot");
+
+    // TODO: this is a repetition of versions from elsewhere
+    Command::new("python3")
+        .args(&["-m", "pip", "install", "cbmc-viewer==2.11", "--target"])
```

Comment:
Have you considered having one source of truth for these dependencies? This looks a bit fragile.

Reply:
Yeah, we could really use a solution to that problem. I haven't yet figured out how we can change dev environments and CI around to use something more like a release bundle approach (which I think would be the best way to have a single source of truth: right here).